### DISCUSSION TITLE
How do I run the tests in goodreads?

### DISCUSSION THREAD

**Discussion by steinarb** [bookwyrm-social/bookwyrm#1679] [Q&A]
I don't have goodreads account to export bookshelves from.

Therefore I am writing this project https://github.com/steinarb/bokbase which is a simple book database that creates goodreads.csv files that can be imported into bookwyrm.

Unfortunately the import of CSV files produced by bokbase into bookwyrm fails.

So then my idea was to replace the goodreads.csv in bookwyrm/bookwyrm/test/data/ with my own file, and use that to create the correct format for bookwyrm.

Unfortunately running the test is hard. After installing all dependencies required for the tests, I still get the following error:
```
bookwyrm/tests/views/preferences/test_delete_user.py:10: in <module>
    from bookwyrm import forms, models, views
bookwyrm/forms.py:11: in <module>
    from bookwyrm import models
bookwyrm/models/__init__.py:5: in <module>
    from .book import Book, Work, Edition, BookDataModel
bookwyrm/models/book.py:12: in <module>
    from imagekit.models import ImageSpecField
/usr/lib/python3/dist-packages/imagekit/__init__.py:2: in <module>
    from . import conf
/usr/lib/python3/dist-packages/imagekit/conf.py:6: in <module>
    class ImageKitConf(AppConf):
/usr/lib/python3/dist-packages/appconf/base.py:75: in __new__
    new_class._configure()
/usr/lib/python3/dist-packages/appconf/base.py:101: in _configure
    value = getattr(obj._meta.holder, prefixed_name, default_value)
/usr/local/lib/python3.9/dist-packages/django/conf/__init__.py:84: in __getattr__
    self._setup(name)
/usr/local/lib/python3.9/dist-packages/django/conf/__init__.py:65: in _setup
    raise ImproperlyConfigured(
E   django.core.exceptions.ImproperlyConfigured: Requested setting IMAGEKIT_CACHEFILE_NAMER, but settings are not configured. You must either define the environment variable DJANGO_SETTINGS_MODULE or call settings.configure() before accessing settings.
``` 

Any ideas what I need to do to be able to run "pytest" without errors?

Thanks!

**Comment by mouse-reeve** [comment]
It looks like there's a missing dependency -- what command did you use to run the test?

**Comment by steinarb** [comment]
>>>>> Mouse Reeve ***@***.***>:

> It looks like there's a missing dependency -- what command did you use to run the test?

I have tried two:
 pytest
and
 python -m pytest

Both give the same error messages.

This is on a debian stable with some python dependencies installed with
apt and some (not found as debian packages), installed with pip3 run as
root.

I have used apt or pip3 to fill a lot of missing dependencies warnings.

**Reply by mouse-reeve** [reply]
are you able to run the tests in docker? the command is `./bw-dev pytest` 

The problem seems to be with `django-imagekit`, which should be installed with pip via the `requirements.txt` file, but it could be that it also needs a package installed to work on debian. With docker, which I recommend if it's possible, the dependencies should be handled automatically.

**Reply by steinarb** [reply]
The command to run tests in docker fails:
```
sb@lorenzo:~/git/bookwyrm$ ./bw-dev pytest
./bw-dev: line 12: .env: No such file or directory
Failed to load configuration! You may need to update your .env and quote values with special characters in them.
sb@lorenzo:~/git/bookwyrm$
```

I am trying the command from an amd64 computer, running debian 11.1 "bullseye", with docker container 20.10.5 installed and running.

I think django-imagekit was the stuff I couldn't find .deb packages for, and installed with pip3 (but the command history has been lost, so I can't be sure).

**Reply by steinarb** [reply]
Found the command history for the installs (they were on the root user):
```
root@lorenzo:~# apt install python3-pytest
root@lorenzo:~# apt install python3-pytest-cov
root@lorenzo:~# pip3 install Django
root@lorenzo:~# apt install python3-psycopg2
root@lorenzo:~# pip3 install django-model-utils
root@lorenzo:~# apt install python3-responses
root@lorenzo:~# pip3 install pytz
root@lorenzo:~# apt install python3-django-imagekit
```
(I was led by google matches and debian package search matches on the missing dependencies I saw)

Maybe the problem is my manual pip3 installs followed by the apt install of python3-django-imagekit?

Is there a way to clean up things?  I'm not really familiar with python (which is somewhat strange, considering my most popular git repo is written in python...).

**Reply by mouse-reeve** [reply]
The docker error looks easier to fix than the one you're dealing with -- it sounds like might be able to resolve it by putting quotes around the values in the .env file. 

I'd recommend using a virtual environment for doing pip installs (https://docs.python.org/3/library/venv.html) to avoid having to install system-wide. It might be worth a shot to re-install via pip after doing the apt install.

**Reply by steinarb** [reply]
The problem was a missing .env. I copied .env.dev.example to .env and then tried again. A new error, but a different one:
```
sb@lorenzo:~/git/bookwyrm$ cp .env.dev.example .env
sb@lorenzo:~/git/bookwyrm$ ./bw-dev pytest
+ case "$CMD" in
+ execweb pytest --no-cov-on-fail
+ docker-compose exec web pytest --no-cov-on-fail
ERROR: No container found for web_1
sb@lorenzo:~/git/bookwyrm$
```

I'm guessing the next possible question is "is the docker container runtime running?", and as far as I can tell, it is, but no containers are currently running:
```
sb@lorenzo:~/git/bookwyrm$ docker ps
CONTAINER ID   IMAGE     COMMAND   CREATED   STATUS    PORTS     NAMES
sb@lorenzo:~/git/bookwyrm$
```

**Reply by mouse-reeve** [reply]
Ah I wrote the bw-dev command to assume that the web image is already running, try `docker-compose run --rm web pytest`

**Reply by steinarb** [reply]
Ok, that did a lot of stuff: First it downloaded several images (postgres and python passed by), and then it stopped in a prompt, and I thought I was back in bash and pressed key up, and the prompt filled with control characters.

And then I pressed RET and Ctrl-C (probably) and then it ran the tests and exited.

I tried starting the tests again with your bw-dev command but that didn't work. 

Here's the output from the end of the test run, followed by "docker ps" and a new attempt at running bw-dev:
```
...
bookwyrm/views/preferences/block.py                   24     10    58%
bookwyrm/views/preferences/change_password.py         20     10    50%
bookwyrm/views/preferences/delete_user.py             23     12    48%
bookwyrm/views/preferences/edit_user.py               45     28    38%
bookwyrm/views/reading.py                             84     56    33%
bookwyrm/views/rss_feed.py                            24     11    54%
bookwyrm/views/search.py                              50     31    38%
bookwyrm/views/shelf/__init__.py                       0      0   100%
bookwyrm/views/shelf/shelf.py                         62     43    31%
bookwyrm/views/shelf/shelf_actions.py                 50     32    36%
bookwyrm/views/status.py                             163    124    24%
bookwyrm/views/updates.py                             12      5    58%
bookwyrm/views/user.py                                67     44    34%
bookwyrm/views/wellknown.py                           49     26    47%
bookwyrm/wsgi.py                                       6      6     0%
----------------------------------------------------------------------
TOTAL                                               7864   4626    41%


!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! KeyboardInterrupt !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
/usr/local/lib/python3.9/site-packages/django/db/models/options.py:484: KeyboardInterrupt
(to show a full traceback on KeyboardInterrupt use --full-trace)
=================================================================== no tests ran in 37.01s ===================================================================
sb@lorenzo:~/git/bookwyrm$ docker ps
CONTAINER ID   IMAGE                    COMMAND                  CREATED              STATUS              PORTS                    NAMES
7869608d3c7d   bookwyrm_celery_worker   "celery -A celerywyr…"   About a minute ago   Up 57 seconds                                bookwyrm_celery_worker_1
48f2e6bef852   postgres                 "docker-entrypoint.s…"   About a minute ago   Up About a minute   5432/tcp                 bookwyrm_db_1
38db47816fa8   redis                    "docker-entrypoint.s…"   About a minute ago   Up About a minute   6379/tcp                 bookwyrm_redis_activity_1
600faea7f0ca   redis                    "docker-entrypoint.s…"   About a minute ago   Up About a minute   0.0.0.0:6379->6379/tcp   bookwyrm_redis_broker_1
sb@lorenzo:~/git/bookwyrm$ ./bw-dev pytest
+ case "$CMD" in
+ execweb pytest --no-cov-on-fail
+ docker-compose exec web pytest --no-cov-on-fail
ERROR: No container found for web_1
sb@lorenzo:~/git/bookwyrm$
```

**Reply by mouse-reeve** [reply]
I'm not sure what the prompt was -- did you get any indication about what triggered it? Is it possible that there was a pdb statement added somewhere in the tests?

You have two options for running the tests, one is to use the `docker-compose run --rm web pytest`, which doesn't require the `web` image to already be running, and, as you saw, will spin up the related images for the database and celery and so on. The only real downside to this is that sometimes externalities that should cause errors in the test, like calls to redis, will work fine because the redis image is available. So if you're writing tests and they work locally but don't work in the github ci, for example, this is often why. If you're just running existing tests, this is a perfectly fine way to do it.

To use the `./bw-dev` command, you'll need the `web` image to be running. You can bring it up with `docker-compose restart web`. When I'm writing mocks for celery and redis, I have the `web` image running, and stop the celery_worker, redis_broker, and redis_activity images.

I know that's long and complicated and may be more info than you need! The shortest version is, the `docker-compose run --rm web pytest` command is the simplest way to run the tests, and I'd be happy to help figure out what the prompt you're hitting is.

**Reply by steinarb** [reply]
> I'm not sure what the prompt was -- did you get any indication about what triggered it? Is it possible that there was a pdb statement added somewhere in the tests?

My mistake! It was the line of the first test I took to be a shell prompt and started typing, ie.:
```
sb@lorenzo:~/git/bookwyrm$ docker-compose run --rm web pytest
Starting bookwyrm_db_1           ... done
Starting bookwyrm_redis_broker_1   ... done
Starting bookwyrm_redis_activity_1 ... done
Starting bookwyrm_celery_worker_1  ... done
==================================================================== test session starts =====================================================================
platform linux -- Python 3.9.9, pytest-6.1.2, py-1.11.0, pluggy-0.13.1
django: settings: bookwyrm.settings (from ini)
rootdir: /app, configfile: pytest.ini
plugins: celery-4.4.2, forked-1.4.0, cov-2.10.1, env-0.6.2, django-4.1.0, xdist-2.3.0
collected 704 items

bookwyrm/tests/test_book_search.py
```

This time I waited and after a while the tests started running.

Is there a way to run just the goodreads import test?  

That's the only one I really want to run, on my own file to see where it goes wrong, and then change my exporter until I produce something bookwyrm accepts.

**Reply by mouse-reeve** [reply]
Ah yeah the tests are very very slow when you run all of them; you can include the path to whatever test file you want to run: `docker-compose run --rm pytest bookwyrm/tests/importers/test_goodreads_import.py` (you can also just include the directory, and it will run all test files in that directory). If you want to specify a specific test within the file, use the `-k` flag, for example: `docker-compose run --rm pytest bookwyrm/tests/importers/test_goodreads_import.py -k test_create_job`

**Reply by steinarb** [reply]
Thanks, this command worked (added a "web" between "--rm" and "pytest"):
```
docker-compose run --rm web pytest bookwyrm/tests/importers/test_goodreads_import.py
```

Where are the file in the test picked up from? 

I.e. if I copy my own file into bookwyrm/test/data/goodreads.csv and re-run the command, will the test be run with my csv file or the one already in the doker image?

Also, is there a way to see the console output from the test run?

Thanks again!

**Reply by mouse-reeve** [reply]
oh yes sorry! it should definitely have web in it. The tests open the files manually using relative paths, in the goodreads tests, it's `datafile = pathlib.Path(__file__).parent.joinpath("../data/goodreads.csv")`, so yes, replacing the goodreads.csv file should mean that it's working from the new file. 

The console output seems to only show up when the tests fail, and is suppressed when they pass. There is probably a way around that, but I haven't found it; I usually just add `raise Exception` to the end of the test so that the test fails, which is janky but does get me the print statements. You can also add pdb statements to the test if you prefer a proper debugger.

**Reply by steinarb** [reply]
Hm... the tests fail on my .csv file, but the print statements don't appear in the output. 

What's output for each failing test, is the code of the failing test, with a ">" pointing at the failing line, e.g. like so:
```
______________________________________________________________ GoodreadsImport.test_create_job _______________________________________________________________

self = <bookwyrm.tests.importers.test_goodreads_import.GoodreadsImport testMethod=test_create_job>
_ = (<MagicMock name='delay' id='140603341207392'>, <MagicMock name='delay' id='140603338347040'>, <MagicMock name='delay' id='140603338503504'>)
import_job = <ImportJob: ImportJob object (1)>, import_items = <QuerySet []>

    def test_create_job(self, *_):
        """creates the import job entry and checks csv"""
        import_job = self.importer.create_job(
            self.local_user, self.csv, False, "public"
        )

        import_items = models.ImportItem.objects.filter(job=import_job).all()
>       self.assertEqual(len(import_items), 3)
E       AssertionError: 0 != 3
```

**Reply by steinarb** [reply]
The reason nothing was output, was that the code I had put the debug trace in was never called, and the reason that code was never called, was that my test file had 0 rows in it.

So: my bad!

To get that far I discovered that [adding the -s flag to pytest, would output all debug trace, interleaved with the pytest output](https://stackoverflow.com/a/14414325):
```
docker-compose run --rm web pytest -s bookwyrm/tests/importers/test_goodreads_import.py
```

And then I traced slowly from the test into the code until I came to the point that failed.

**Reply by mouse-reeve** [reply]
I see - thank you for debugging, I was totally flummoxed as to why it wouldn't be printing
